src/Workspaces/Core/Portable/ReassignedVariable/AbstractReassignedVariableService.cs

```diff
@@ -42,112 +42,136 @@ internal abstract class AbstractReassignedVariableService<
             var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
             var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
 
+            var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
 
             using var _1 = PooledDictionary<ISymbol, bool>.GetInstance(out var symbolToIsReassigned);
             using var _2 = ArrayBuilder<TextSpan>.GetInstance(out var result);
-            using var _3 = ArrayBuilder<SyntaxNode>.GetInstance(out var stack);
+            using var _3 = PooledDictionary<SyntaxTree, SemanticModel>.GetInstance(out var syntaxTreeToModel);
 
-            // Walk through all the nodes in the provided span.  Directly analyze local or parameter declaration.  And
-            // also analyze any identifiers which might be reference to locals or parameters.  Note that we might hit
-            // locals/parameters without any references in the span, or references that don't have the declarations in 
-            // the span
-            stack.Add(root.FindNode(span));
+            return Recurse();
```

Comment:
pulled into a nested function so that we wouldn't scope 'semanticModel' over all helper functions below (since we now have to analyze multiple partial types whihc may be in other syntax trees.